tools/chpl-language-server/src/chpl-language-server.py

```diff
@@ -628,73 +670,124 @@ def _enter_Module(self, node: chapel.Module):
+                return None
 
-                # Only show nodes without @chpldoc.nodoc. The exception
-                # about standard files applies here too.
-                documented_nodes = []
-                for node in nodes:
-                    # apply aforementioned exception
-                    if in_bundled_module:
-                        documented_nodes.append(node)
-                        continue
+            # Just take the first value to avoid showing N entries for
+            # overloaded functions.
+            return name, documented_nodes[0]
 
-                    # avoid nodes with nodoc attribute.
-                    ag = node.attribute_group()
-                    show = False
-                    if not ag or not ag.get_attribute_named("chpldoc.nodoc"):
-                        show = True
-                    elif name in _ALLOWED_NODOC_DECLS:
-                        # If users declare variables like 'here' themselves,
-                        # we will not show them if they're @chpldoc.nodoc,
-                        # since they're not special.
-                        decl_file = node.location().path()
-                        is_standard_decl = self.context.context.is_bundled_path(
-                            decl_file
+        def visible_nodes_for_scopes(
+            node: chapel.AstNode, scopes: List[chapel.Scope]
+        ):
+            visible_nodes = []
+            file = node.location().path()
+            in_bundled_module = self.context.context.is_bundled_path(file)
+            for depth, scope in enumerate(scopes):
+                files_named_in_use_or_import = set()
+                for m in scope.modules_named_in_use_or_import():
+                    files_named_in_use_or_import.add(m.location().path())
+
+                for name, nodes in scope.visible_nodes():
+                    visible_node = visible_nodes_for_scope(
+                        name, nodes, in_bundled_module
+                    )
+                    if visible_node:
+                        d = depth
+                        visible_path = visible_node[1].location().path()
+                        if visible_path != file:
+                            # if from a different file, increase the depth by 1
+                            d += 1
+                            # if from a bundled path increase the depth by 1
+                            d += int(
+                                self.context.context.is_bundled_path(
+                                    visible_path
+                                )
+                            )
+                            # if not explicitly used, increase the depth by 1
+                            d += int(
+                                visible_path not in files_named_in_use_or_import
+                            )
+
+                        visible_nodes.append(
+                            (visible_node[0], visible_node[1], d)
                         )
```

Comment:
This is getting really deep and long, can you flatten and split out helper functions?